Optimizations of the HGCAL L1 trigger

jbsauvan opened the pull request:
This pull request includes the following changes to the HGCAL L1 trigger:
- Optimization of the processing time of the HGCAL trigger. This is mainly a reorganization of loops in the front-end part of the trigger.
- Optimization of the memory footprint of the HGCAL trigger geometry. It now makes use of the detid structure to associate cells to trigger cells and trigger modules, instead of storing the full trigger cells and modules maps.  This involves changes in the interface of the geometry.
- Updates in the HGCAL trigger tests
- Some bugfixes

baffioni committed:
Merged hgc-tpg-devel-CMSSW_8_1_0_pre6 from repository PFCal-dev

baffioni committed:
add -1 to the saturation value

baffioni committed:
association with simhits

baffioni committed:
association with simhits

baffioni committed:
SimDataFormats/CaloTest package added

baffioni committed:
corrections after comments

baffioni committed:
corrections

jbsauvan committed:
Merge pull request #35 from baffioni/baffioni-smihits-810pre6

Baffioni smihits 810pre6

jbsauvan committed:
Identify trigger cells within modules instead of wafers

jbsauvan committed:
Merged hgc-tpg-devel-CMSSW_8_1_0_pre6 from repository PFCal-dev

jbsauvan committed:
Update configs to pre9 release

jbsauvan committed:
Merge branch 'hgc-tpg-devel-CMSSW_8_1_0_pre9' into fix-triggercell-sums

jbsauvan committed:
Merge pull request #36 from jbsauvan/fix-triggercell-sums

Update trigger cell id

jbsauvan committed:
Improve front-end processing time

jbsauvan committed:
Remove virtual keywords

jbsauvan committed:
Replace insert with emplace

jbsauvan committed:
Replace maps with unordered_maps

jbsauvan committed:
Revert back to ordered maps in some cases

jbsauvan committed:
Merge updates from 81X PR

jbsauvan committed:
Merge pull request #37 from jbsauvan/optimize-frontend

Optimize front-end processing time

jbsauvan committed:
Merge pull request #38 from jbsauvan/merge-81x-pr-updates

81X PR updates

jbsauvan committed:
Add new trigger detid

jbsauvan committed:
Add lightweight geometry class

jbsauvan committed:
Update module drawing for the lightweight geometry

jbsauvan committed:
Clean code related to lightweight geometry

jbsauvan committed:
Add test class for best choice timing study

jbsauvan committed:
Change if/else to switch/case

jbsauvan committed:
Make HGCTriggerHexDetId format derive from HGCalDetId

jbsauvan committed:
Add HGCTriggerHexDetId in classes_def

jbsauvan committed:
Add position methods in geometry

jbsauvan committed:
Remove getValidIds methods

jbsauvan committed:
Use new geometry interface

jbsauvan committed:
Cleaning

jbsauvan committed:
Change HGCalDetId to HGCTriggerHexDetId

jbsauvan committed:
Merge pull request #39 from jbsauvan/reduce-geometry-memory

Reduce geometry memory

jbsauvan committed:
Fix trigger cell sums

jbsauvan committed:
Update trigger cell energy truncation

jbsauvan committed:
Fix comment

jbsauvan committed:
Merge pull request #40 from jbsauvan/change-bit-truncation

Change bit truncation

jbsauvan committed:
Merging hgc-tpg-devel-CMSSW_8_1_0_pre9 in CMSSW_8_1_0_pre11

jbsauvan committed:
Update geometry config

jbsauvan committed:
Add dictionaries for vectors of L1THGCal objects

jbsauvan committed:
Merged hgc-tpg-devel-CMSSW_8_1_0_pre11 from repository PFCal-dev

cmsbuild commented:
A new Pull Request was created by  @jbsauvan (Jean-Baptiste Sauvan) for CMSSW_8_1_X.

It involves the following packages:

DataFormats/ForwardDetId
DataFormats/L1THGCal
L1Trigger/L1THGCal


@civanch, @mdhildreth, @cmsbuild, @rekovic, @mulhearn, @davidlange6 can you please review it and eventually sign? Thanks.
@Martin-Grunewald, @amarini, @lgray this is something you requested to watch as well.
@slava77, @smuzaffar you are the release manager for this.

cms-bot commands are list here cms-sw/cmssw#13028

lgray commented:
@jbsauvan Is the L1 for HGCal run in the phase2 workflows yet? (or does it have any reason to be run, even in part?) Just asking for the purposes of automated test, etc.

lgray commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

jbsauvan commented:
@lgray No, currently it is not included in any phase2 workflows. So the only automated tests affected by the HGCAL L1 are the unit tests in L1THGCal.
But I am starting now to work on integrating the HGCal L1 trigger in the phase2_hgcal era.

cmsbuild commented:
+1
Tested at: 9d840ce4a505435ff392a16b591f651336f1d062
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-15782/15052/summary.html

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/9d840ce4a505435ff392a16b591f651336f1d062
https://github.com/cms-sw/cmssw/commit/33b90280ffcb2fdbbe73e7d85c978a4c2dbbbe5c
https://github.com/cms-sw/cmssw/commit/864910059e29918db17cc0a7ffae4f9aa0bd2ea6
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-15782/15052/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-15782/15052/git-merge-result

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-15782/15052/summary.html

civanch commented:
+1

davidlange6 commented on DataFormats/ForwardDetId/interface/HGCTriggerHexDetId.h:
```diff
@@ -0,0 +1,78 @@
+#include <iosfwd>
+#include "DataFormats/ForwardDetId/interface/HGCalDetId.h"
+#include "DataFormats/ForwardDetId/interface/ForwardSubdetector.h"
+#include "TObject.h"
```
@jbsauvan - why is include this needed?

jbsauvan replied:
@davidlange6 
This is not needed, I am removing it.

davidlange6 commented:
@jbsauvan - two more generic questions - why have a separate DetId class that differs only by static functions from existing ones? Or do I miss some differences? Otherwise the *Old classes - what does "Old" mean exactly? Usually more descriptive names are useful (or is this just obsolete stuff?)

jbsauvan committed:
Remove TObject include in HGCTrigger detids

cmsbuild commented:
Pull request #15782 was updated. @civanch, @mdhildreth, @cmsbuild, @rekovic, @mulhearn, @davidlange6 can you please check and sign again.

jbsauvan commented:
@davidlange6 
* Concerning the DetId class, the goal was to keep the HGCalDetId unchanged,  such that it doesn't affect non-trigger parts. For the moment the only differences are static functions, but there may be more differences added in the future. If you think it can be done in some better ways, I would be happy to change this.
* Concerning the "Old" geometry class, it is only there for backward compatibility, but I agree the name should be more descriptive; I am changing it.

davidlange6 commented:
> On Sep 12, 2016, at 2:36 PM, Jean-Baptiste Sauvan <notifications@github.com> wrote:
> 
> @davidlange6
> 
> 	• Concerning the DetId class, the goal was to keep the HGCalDetId unchanged, such that it doesn't affect non-trigger parts. For the moment the only differences are static functions, but there may be more differences added in the future. If you think it can be done in some better ways, I would be happy to change this.

Seems it can either be just utility functions or in the HGCalDetId class - I don’t see anything in these static functions that seems to be trigger specific

> 	• Concerning the "Old" geometry class, it is only there for backward compatibility, but I agree the name should be more descriptive; I am changing it.

backwards compatibility in which sense?

Cheers-
David

> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

jbsauvan commented:
@davidlange6 
> Seems it can either be just utility functions or in the HGCalDetId class - I don’t see anything in these static functions that seems to be trigger specific

Ok. So if I change the HGCalDetId class adding these static functions, will it be possible to read HGCalDetId objects from files with the previous class version without any problem?

> backwards compatibility in which sense?

I'm not sure if "backward compatible" is the right term. This *Old geometry class makes possible to still use the old geometry concrete classes even if the interface of the base class has changed. These old geometry classes are not supposed to be used, but just in case it is possible to use them.

jbsauvan committed:
Rename HGCalTriggerGeometryOld to HGCalTriggerGeometryGenericMapping

cmsbuild commented:
Pull request #15782 was updated. @civanch, @mdhildreth, @cmsbuild, @rekovic, @mulhearn, @davidlange6 can you please check and sign again.

jbsauvan committed:
Use HGCalDetId instead of HGCTriggerHexDetId

cmsbuild commented:
Pull request #15782 was updated. @civanch, @mdhildreth, @cmsbuild, @rekovic, @mulhearn, @davidlange6 can you please check and sign again.

jbsauvan commented:
@davidlange6 
I have removed the HGCTriggerHexDetId class, and added some [utility functions](https://github.com/PFCal-dev/cmssw/blob/16bb4977d72d694e15622206326ece28acb178c3/DataFormats/ForwardDetId/interface/HGCalDetId.h#L62-L81) in the HGCalDetId files.

Cheers,
Jean-Baptiste

jbsauvan commented:
@rekovic @mulhearn 
Could you please have a look at this PR and comment or sign?

jbsauvan commented:
And could someone trigger the tests again?
Thanks

lgray commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 16bb4977d72d694e15622206326ece28acb178c3
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-15782/15136/summary.html

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-15782/15136/summary.html

civanch commented:
+1

jbsauvan commented:
@rekovic @mulhearn 
Could you please have a look at this PR?
If you need more information (e.g. a presentation), please tell me.
thanks

rekovic commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next CMSSW_8_1_X IBs (tests are also fine). This pull request requires discussion in the ORP meeting before it's merged. @slava77, @davidlange6, @smuzaffar

davidlange6 commented on DataFormats/ForwardDetId/interface/HGCalDetId.h:
```diff
@@ -59,6 +59,27 @@ class HGCalDetId : public DetId {
+  int zsideOf(uint32_t id);
+
+  void setCellOf(uint32_t& id, int cell);
+  void setWaferOf(uint32_t& id, int mod);
```
@jbsauvan - looking again at these changes and how you use the functions defined - I don't see why you don't just use HGCalDetId instead of the uint32_t in your functions below. There should be no overhead in doing that (and it seems all the functionality needed is already there.

lgray replied:
@jbsauvan @davidlange6 Actually if you are passing uint32_t you should pass by value (32 bits wide), rather than by reference (64 bits wide on most systems nowadays).

lgray replied:
And, consequently have this function return the new detid, or simply forcibly inline this function.

davidlange6 commented:
> On Sep 16, 2016, at 12:42 PM, Lindsey Gray <notifications@github.com> wrote:
> 
> @lgray commented on this pull request.
> 
> In DataFormats/ForwardDetId/interface/HGCalDetId.h:
> 
> > @@ -59,6 +59,27 @@ class HGCalDetId : public DetId {
>    static const HGCalDetId Undefined;
>  };
>  
> +namespace HGCalDetIdUtil
> +{
> +  // Functions to work on raw HGCAlDetIds
> +  int subdetIdOf(uint32_t id);
> +  int cellOf(uint32_t id);
> +  int waferOf(uint32_t id);
> +  int waferTypeOf(uint32_t id);
> +  int layerOf(uint32_t id);
> +  int zsideOf(uint32_t id);
> +
> +  void setCellOf(uint32_t& id, int cell);
> +  void setWaferOf(uint32_t& id, int mod);
> 
> @jbsauvan @davidlange6 Actually if you are passing uint32_t you should pass by value (32 bits wide), rather than by reference (64 bits wide on most systems nowadays).

id gets updated inside these functions…

(anyway, I would prefer all this be replaced by the DetId itself.)

> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

jbsauvan commented:
@davidlange6 @lgray 
Ok. So what I can do is to keep HGCalDetId as it is, remove all these additional functions and use HGCalDetId objects instead of calling these utility functions, like for instance in the trigger geometry class have things like:
```python
unsigned 
HGCalTriggerGeometryHexImp2::
getTriggerCellFromCell( const unsigned cell_det_id ) const
{
    HGCalDetId cell_id(cell_det_id);
    int wafer_type = cell_id.waferType();
    unsigned cell = cell_id.cell();
    unsigned trigger_cell = cells_to_trigger_cells_.at(std::make_pair(wafer_type,cell));
    HGCalDetId trigger_cell_det_id((ForwardSubdetector)cell_id.subdetId(), cell_id.zside(), cell_id.layer(), cell_id.waferType(), cell_id.wafer(), trigger_cell);
    return trigger_cell_det_id.rawId();
}
```
instead of 
```python
unsigned 
HGCalTriggerGeometryHexImp2::
getTriggerCellFromCell( const unsigned cell_det_id ) const
{
    int wafer_type = HGCalDetIdUtil::waferTypeOf(cell_det_id);
    unsigned cell = HGCalDetIdUtil::cellOf(cell_det_id);
    unsigned trigger_cell = cells_to_trigger_cells_.at(std::make_pair(wafer_type,cell));
    unsigned trigger_cell_det_id = cell_det_id;
    HGCalDetIdUtil::setCellOf(trigger_cell_det_id, trigger_cell);
    return trigger_cell_det_id;
}
```
I checked that it is exactly the same in terms of (processing time) performance.

If this is ok for you, I will push the changes.

lgray commented:
@jbsauvan Make it nameless:
```
unsigned 
HGCalTriggerGeometryHexImp2::
getTriggerCellFromCell( const unsigned cell_det_id ) const
{
    HGCalDetId cell_id(cell_det_id);
    int wafer_type = cell_id.waferType();
    unsigned cell = cell_id.cell();
    unsigned trigger_cell = cells_to_trigger_cells_.at(std::make_pair(wafer_type,cell));
    return HGCalDetId((ForwardSubdetector)cell_id.subdetId(), cell_id.zside(), cell_id.layer(), wafer_type, cell_id.wafer(), trigger_cell).rawId();
}
```

jbsauvan committed:
Remove HGCalDetId utilities

cmsbuild commented:
Pull request #15782 was updated. @civanch, @mdhildreth, @cmsbuild, @rekovic, @mulhearn, @davidlange6 can you please check and sign again.

jbsauvan commented:
@davidlange6 @lgray 
I have updated the code according to the comments above.

lgray commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: f8b46dbad084114c774f6bdbba2bf21b8b923263
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-15782/15240/summary.html

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-15782/15240/summary.html

jbsauvan commented:
@davidlange6 
Is the updated code ok for you? Is there something else that could be improved?
Thanks

jbsauvan commented:
@civanch @mdhildreth @mulhearn @rekovic Could you check if the updated PR is still ok? (And sign again if this is the case)
@davidlange6 Could you also sign if you are satisfied with the changes?
Thanks

jbsauvan commented:
@civanch @mdhildreth @mulhearn @rekovic @davidlange6 
If you have no more comments and are satisfied with the changes, could you please consider signing (again, for those who already signed) and merging this PR.
I really would like this to be included in pre13 since it brings significant performance improvement.
Thanks
